Sources/MongoSwift/ServerSelection.swift

```diff
@@ -0,0 +1,308 @@
+                    )
+                }
+            }
+
+            var suitableServers = try topology.findSuitableServers(
+                readPreference: readPreference,
+                heartbeatFrequencyMS: self.connectionString.heartbeatFrequencyMS
+                    ?? SDAMConstants.defaultHeartbeatFrequencyMS
+            )
+            suitableServers.filterByLatency(localThresholdMS: self.connectionString.localThresholdMS)
+            var inWindowServers = suitableServers.compactMap { servers[$0.address] }
+
+            let selectedServer: Server
+            if inWindowServers.isEmpty {
+                // When pure Swift SDAM is implemented, this should instead block on a topology change occurring for
+                // endTime - Date() seconds.
+                // TODO: add an async sleep here once the driver has been updated to permit async/await on 10.15+
+                continue
+            } else if inWindowServers.count == 1 {
+                selectedServer = inWindowServers[0]
+            } else {
+                // Remove the randomly chosen servers from the list to ensure that the same server is not chosen twice.
+                let server1 = inWindowServers.remove(at: Int.random(in: 0..<inWindowServers.count))
+                let server2 = inWindowServers.remove(at: Int.random(in: 0..<inWindowServers.count))
+
+                if server1.operationCount.load(ordering: .sequentiallyConsistent)
+                    <= server2.operationCount.load(ordering: .sequentiallyConsistent)
+                {
+                    selectedServer = server1
+                } else {
+                    selectedServer = server2
+                }
+            }
+
+            selectedServer.operationCount.wrappingIncrement(ordering: .sequentiallyConsistent)
+            return selectedServer
+        }
+
+        throw MongoError.ServerSelectionError(message: "Server selection timed out: no suitable servers found using"
+            + " read preference: \(readPreference)\nTopology: \(topology)")
+    }
+}
+
+extension Array where Element == ServerDescription {
+    /// Filters servers according to their latency. A server is considered to be within the latency window if its
+    /// `averageRoundTripTimeMS` is no more than `localThresholdMS` (or 15 by default) milliseconds greater than the
+    /// smallest average RTT seen amongst the servers.
+    internal mutating func filterByLatency(localThresholdMS: Int?) {
+        guard let minAverageRoundTripTime = self.compactMap({ $0.averageRoundTripTimeMS }).min() else {
+            // If there is no minimum average round trip time, there are no servers to filter.
+            return
+        }
+        let maxAverageRoundTripTime = minAverageRoundTripTime
+            + Double(localThresholdMS ?? SDAMConstants.defaultLocalThresholdMS)
+        self.removeAll {
+            guard let averageRoundTripTimeMS = $0.averageRoundTripTimeMS else {
```

Comment:
per the comment above, this should never really happen right?
it might be good to leave a comment in the code somewhere in this method about this just so we remember when looking at this logic in the future